modules/dag/src/main/scala/org/tessellation/dag/l1/DAGBlockValidator.scala

```diff
@@ -0,0 +1,178 @@
+//  def containsAcceptedOrStaleTx(cbd: DAGBlockData): F[List[String]] = {
+//    val containsAccepted = cbd.transactions.toList.map { t =>
+//
+//      transactionStorage.getLastAcceptedTransactionRef(t.baseHash).map {
+//        //b => (t.hash, b)
+//        case Some(tr) =>
+//        case None =>
+//      }
+//    }.sequence[F, (String, Boolean)]
+//      .map(l => l.collect { case (h, true) => h })
+//
+//    containsAccepted
+//  }
+
+  private def areTransactionsValid(signedTransactions: Seq[Signed[Transaction]]): F[Boolean] =
+    for {
+      arePassingValidation <- areTransactionsPassingValidation(signedTransactions)
+      areChainedCorrectly <- areAllTransactionsFormingCorrectChain(signedTransactions)
+    } yield arePassingValidation && areChainedCorrectly
+
+  private def areTransactionsPassingValidation(signedTransactions: Seq[Signed[Transaction]]): F[Boolean] =
+    transactionValidator
+      .validateBatchOfTransactions(signedTransactions)
+      .map(_.isValid)
+      .handleError(_ => false)
+
+  private def areAllTransactionsFormingCorrectChain(transactions: Seq[Signed[Transaction]]): F[Boolean] =
+    transactions
+      .groupBy(_.value.source)
+      .toList
+      .traverse {
+        case (address, txs) =>
+          for {
+            lastAccepted <- transactionStorage.getLastAcceptedTransactionRef(address)
+            sorted = txs.sortBy(_.value.parent.ordinal.coerce.value)
+            // TODO: do we also need to validate if the first lastRef isn't referencing itself?
+            //  I guess it makes no sense given that ordinal and hash of the current tx isnt specified so it's impossible to self-reference
+            chainableTransactions = takeConsecutiveTransactions(lastAccepted, txs)
+            //  was txs.toSet == chainableTransactions.toSet but that deduplicates, went with sorted lists as these should match
+          } yield sorted == chainableTransactions
+      }
+      .map(_.forall(identity))
+
+  private def prepareBalances(transactions: Seq[Transaction]): F[Seq[UpsertAction[StoredAddress]]] =
+    transactions.traverse { tx =>
+      for {
+        sourceBalance <- addressStorage.getBalance(tx.source)
+        newSourceBalance <- refineV[NonNegative]
+          .apply(sourceBalance.coerce.value - tx.amount.coerce.value)
+          .bimap(new Throwable(_), Balance(_))
+          .liftTo[F]
+        sourceAction = Insert(StoredAddress(tx.source, newSourceBalance))
+        maybeDestinationBalance <- addressStorage.getMaybeBalance(tx.destination)
+        destinationAction = maybeDestinationBalance.map(db => Balance(db.coerce |+| tx.amount.coerce)) match {
+          case Some(newBalance) =>
+            Update(StoredAddress(tx.destination, newBalance))
+          case None =>
+            Insert(StoredAddress(tx.destination, Balance(tx.amount.coerce)))
+        }
```

Comment:
That would be great, but I couldn't pull that of (thought maybe it can be done) to fetch balance, adjust and generate needed action and do it transactionally for a batch of transaction. So I came up with this idea of `UpsertAction` because I could pre-generate the actions based on current balances (this also requires that only block acceptance updates the balances single block at a time). You are right that for the multiple transactions it won't work - good catch - but I think we could still leverage it just by aggregating the balances from all transactions. We can create a ticket for it and tackle that issue again, because currently I don't know what is the perfect solution.